tests/integration/test_lock.py

```diff
@@ -791,3 +791,39 @@ def test_pipenv_respects_package_index_restrictions(PipenvInstance):
                                       'sha256:ec22d826a36ed72a7358ff3fe56cbd4ba69dd7a6718ffd450ff0e9df7a47ce6a'],
                            'index': 'local', 'version': '==2.19.1'}
         assert p.lockfile['default']['requests'] == expected_result
+
+
+@flaky
+@pytest.mark.dev
+@pytest.mark.lock
+@pytest.mark.install
+def test_dev_lock_use_default_packages_as_constraint(PipenvInstance):
```

Comment:
Actually there are 2 constraints files. The first one is inside resolver.py
https://github.com/pypa/pipenv/blob/30a6b1a5b82f0d04be14274b189467f10be2c3a2/pipenv/utils/resolver.py#L591-L595
The second one is `write_constraint_to_file` as mention above.

- The first constraints file is created and used when: Locking or Installing with Pipfile using `--dev` flag (user don't pass any packages args).
- The second constrains file is created and used when: User pass packages as argument using `--dev` flag.

When locking, the code does the following things:
- Resolve dependencies with `venv_resolve_deps` (this create and use the first constraints file).
- Running `batch_install`, install all packages in the lock file. This one calls `pip_install`, this could create the second constraints file (if both `dev` and `use_constraint` equal True). But since all the packages were resolved in step 1 and written to lockfile, so resolving isn't necessary.
https://github.com/pypa/pipenv/blob/30a6b1a5b82f0d04be14274b189467f10be2c3a2/pipenv/core.py#L780-L793

Reply:
Thanks @dqkqd -- I was able to hit that breakpoint in the test `test_install_dev_use_default_constraints` after I posted the first message.   Can you explain more why to have two different methods of similar code for generating the two different constraints files?

Reply:
We have `lock` and `install` method. The first constraint file in `resolver.py` is used for `lock`, the second one in `core.py` is used for `install`.
1. The first one should not be removed, because it would cause lockfile is not generated correctly.
2. The second one could be remove. Actually, this is my first approach.
Suppose I remove this one. Then `pip_install` inside this loop should not be called, because it will install packages without resolving dependencies.
https://github.com/pypa/pipenv/blob/30a6b1a5b82f0d04be14274b189467f10be2c3a2/pipenv/core.py#L2136-L2271
So `pip_install` should be replaced by another `check_installable_package` method. After that, resolving all the packages with `do_lock` (which use the first constraints file) will give us the same result.
I didn't use this approach because:
    - `$ pip search` is not working, I could not find a way to check if a package is installable without calling `$ pip install`
    - The loop locks scary.

Reply:
Also, there is another case where resolving dependencies inside `pip_install` is necessary.
For example, I have packages installed with this Pipfile:
```toml
[packages]
django = "<3.0"
```
Suppose I install new dev package using command `$ pipenv install --dev "django>3.0,<4.0"`. Clearly the package is installable but it shouldn't be, because it would confliect with default packages.